### PR TITLE
Enable train with fixed confidence threshold f1 score metric

### PR DESCRIPTION
**Pull request by jaegukhyun**
### Checklist

- [ ] I have added unit tests to cover my changes.​
- [ ] I have added integration tests to cover my changes.​
- [ ] I have added e2e tests for validation.
- [ ] I have added the description of my changes into CHANGELOG in my target branch (e.g., [CHANGELOG](https://github.com/openvinotoolkit/training_extensions/blob/develop/CHANGELOG.md) in develop).​
- [ ] I have updated the documentation in my target branch accordingly (e.g., [documentation](https://github.com/openvinotoolkit/training_extensions/tree/develop/docs) in develop).
- [ ] I have [linked related issues](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

### License

- [ ] I submit _my code changes_ under the same [Apache License](https://github.com/openvinotoolkit/training_extensions/blob/develop/LICENSE) that covers the project.
      Feel free to contact the maintainers if that's a concern.
- [ ] I have updated the license header for each file (see an example below).

```python
# Copyright (C) 2024 Intel Corporation
# SPDX-License-Identifier: Apache-2.0
```

### CODE REVIEW

**Review by vinnamkim — COMMENTED** [review]
Why this change is needed? I doubt that there exists a real use-case to setup a confidence threshold by hands during train/val.

**Comment by jaegukhyun** [comment]
> Why this change is needed? I doubt that there exists a real use-case to setup a confidence threshold by hands during train/val.

Unfortunately, Geti has parameters for this.

**Comment by vinnamkim** [comment]
> > Why this change is needed? I doubt that there exists a real use-case to setup a confidence threshold by hands during train/val.
> 
> Unfortunately, Geti has parameters for this.

I think that it should be deprecated.

**Comment by jaegukhyun** [comment]
> > > Why this change is needed? I doubt that there exists a real use-case to setup a confidence threshold by hands during train/val.
> > 
> > 
> > Unfortunately, Geti has parameters for this.
> 
> I think that it should be deprecated.

I'm not sure about this. Could you share your thought with other stakeholders?

**Comment by vinnamkim** [comment]
> > > > Why this change is needed? I doubt that there exists a real use-case to setup a confidence threshold by hands during train/val.
> > > 
> > > 
> > > Unfortunately, Geti has parameters for this.
> > 
> > 
> > I think that it should be deprecated.
> 
> I'm not sure about this. Could you share your thought with other stakeholders?

I reported this issue to Wonju.

**Comment by jaegukhyun** [comment]
> > > > > Why this change is needed? I doubt that there exists a real use-case to setup a confidence threshold by hands during train/val.
> > > > 
> > > > 
> > > > Unfortunately, Geti has parameters for this.
> > > 
> > > 
> > > I think that it should be deprecated.
> > 
> > 
> > I'm not sure about this. Could you share your thought with other stakeholders?
> 
> I reported this issue to Wonju.

Any conclusion?

**Comment by vinnamkim** [comment]
> Any conclusion?

Further discussion with Geti stakeholders is needed

**Comment by jaegukhyun** [comment]
> > Any conclusion?
> 
> Further discussion with Geti stakeholders is needed

Then how about adding this for now. At this moment, we need three postprocessing parameters confidence threshold, use ellipse shapes, max num detections. Now I'm considering add this as a postprocessing parameters for det, iseg model. If it is done, we can easily delete unnecessary post processing parameters

File: src/otx/core/model/detection.py [hunk 1]
```diff
@@ -61,6 +62,7 @@ def __init__(
             torch_compile=torch_compile,
         )
         self.tile_config = TileConfig()
+        self.confidence_threshold: float | None = confidence_threshold
```
Comment (vinnamkim) [inline src/otx/core/model/detection.py:65]: Is this already captured by `self.hparams`? Even if not, please store this in `self.hparams` as the same place as `best_confidence_threshold`.

File: src/otx/core/model/detection.py [hunk 2]
```diff
@@ -163,24 +165,23 @@ def load_state_dict(self, ckpt: dict[str, Any], *args, **kwargs) -> None:
...
-                self.hparams["best_confidence_threshold"] = best_confidence_threshold
+            # NOTE: Validation metric logging can update `best_confidence_threshold`,
+            # if there's no pre-defined best confidence threshold
+            if self.confidence_threshold is None:
+                retval = super()._log_metrics(meter, key)
+                if best_confidence_threshold := getattr(meter, "best_confidence_threshold", None):
+                    self.hparams["best_confidence_threshold"] = best_confidence_threshold
+            else:
+                retval = super()._log_metrics(meter, key, best_confidence_threshold=self.confidence_threshold)
+                self.hparams["best_confidence_threshold"] = self.confidence_threshold
```
Comment (vinnamkim) [inline src/otx/core/model/detection.py:176]: According to the naming change,
```suggestion
            # if there's no fixed confidence threshold
            if fixed_confidence_threshold := self.hparam.get("fixed_confidence_threshold", None):
                retval = super()._log_metrics(meter, key, best_confidence_threshold=fixed_confidence_threshold)
                self.hparams["best_confidence_threshold"] = fixed_confidence_threshold
            else:
                retval = super()._log_metrics(meter, key)
                self.hparams["best_confidence_threshold"] = getattr(meter, "best_confidence_threshold", None)
```

File: src/otx/core/model/detection.py (hunk 1 above)
Comment (vinnamkim) [reply src/otx/core/model/detection.py:65]: Can we rename this to `fixed_confidence_threshold` to make it clear?
```suggestion
        self.fixed_confidence_threshold: float | None = fixed_confidence_threshold
```

**Review by vinnamkim — COMMENTED** [review]
> Then how about adding this for now. At this moment, we need three postprocessing parameters confidence threshold, use ellipse shapes, max num detections. Now I'm considering add this as a postprocessing parameters for det, iseg model. If it is done, we can easily delete unnecessary post processing parameters

Sure. However, I left some comments. Please see

File: src/otx/core/model/detection.py (hunk 2 above)
Comment (jaegukhyun) [reply src/otx/core/model/detection.py:176]: This logic will save best_confidence_threshold and fixed_confidence_threshold filed, and the value of both filed would be same if users specify their own threshold. Is it right?

Comment (vinnamkim) [reply src/otx/core/model/detection.py:176]: Hmm, I thought that you wanted to make them same. If not they are equal to each other, we cannot do the test with the fixed confidence threshold because your test metric logic is
https://github.com/openvinotoolkit/training_extensions/blob/01a07f6ecd16db153f5c4d6cd81717f253325ab9/src/otx/core/model/detection.py#L180-L184

Under this implementation, we have to set `best_confidence_threshold = fixed_confidence_threshold`.

Yea, personally, I prefer the two values ​​to be independent each other. If so, we need to change this part too. `fixed_confidence_threshold` should be get prioritized than `best_confidence_threshold` when it exists during the test phase.

Comment (jaegukhyun) [reply src/otx/core/model/detection.py:176]: How about fix hparam's field name as confidence threshold, if user specify fixed confidence threshold then confidence threshold would be fixed confidence threshold if not, it will be best confidence threshold from metric. And test phase will  only take confidence threshold from hparam or xml file

**Comment by wonjuleee** [comment]
> > > Any conclusion?
> > 
> > 
> > Further discussion with Geti stakeholders is needed
> 
> Then how about adding this for now. At this moment, we need three postprocessing parameters confidence threshold, use ellipse shapes, max num detections. Now I'm considering add this as a postprocessing parameters for det, iseg model. If it is done, we can easily delete unnecessary post processing parameters

Hi @jaegukhyun, thank you for your efforts. After discussion with Radwan, it would be good for holding at this moment. He said, it is a wrong behavior (deterministic confidence threshold during training) and should be applicable only for testing/inference. This will be also fixed in Geti side too.

**Comment by jaegukhyun** [comment]
This will be revisited after Geti's direction of this feature is determined.
